pkg/api/validation/validation.go

```diff
@@ -1237,6 +1256,21 @@ func ValidatePersistentVolume(pv *api.PersistentVolume) field.ErrorList {
 			allErrs = append(allErrs, validateScaleIOVolumeSource(pv.Spec.ScaleIO, specPath.Child("scaleIO"))...)
 		}
 	}
+	if pv.Spec.LocalVolume != nil {
+		if numVolumes > 0 {
+			allErrs = append(allErrs, field.Forbidden(specPath.Child("localVolume"), "may not specify more than 1 volume type"))
+		} else {
+			numVolumes++
+			if !utilfeature.DefaultFeatureGate.Enabled(features.PersistentLocalVolumes) {
+				allErrs = append(allErrs, field.Forbidden(specPath.Child("localVolume"), "LocalVolumes are disabled by feature-gate"))
+			}
+			allErrs = append(allErrs, validateLocalVolumeSource(pv.Spec.LocalVolume, specPath.Child("localVolume"))...)
+			// StorageClass is required
+			if len(pv.Spec.StorageClassName) == 0 {
```

Comment:
Trying to understand why storageclass is needed here; say we have a PV created by our 'secondary partition discovery daemon':

```yaml
kind: PersistentVolume
apiVersion: v1
metadata:
  name: local-pv-1
  labels:
    kubernetes.io/hostname: node-1
spec:
  capacity:
    storage: 100Gi
  localStorage:
    fs:
      path: /var/lib/kubelet/storage-partitions/local-pv-1
  accessModes:
    - ReadWriteOnce
  persistentVolumeReclaimPolicy: Delete
  storageClassName: local-fast
```

Following is the storageclass we have:
```yaml
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: local-fast
provisioner: ""
toplogyKey: kubernetes.io/hostname
```

Now user creates a PVC to claim a local volume:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: myclaim
spec:
  storageClaasName: local-fast
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 8Gi
```

PV/PVC controller will not bind the PVC; rather, when user creates a Pod:

```yaml
kind: Pod
apiVersion: v1
metadata:
  name: mypod
spec:
  containers:
    - name: localvolume-pod
      image: nginx
      volumeMounts:
      - mountPath: "/var/www/html"
        name: mypd
  volumes:
    - name: mypd
      persistentVolumeClaim:
        claimName: myclaim
```

At this moment, scheduler will get the PVC `myclaim`, find suitable PVs, then make a desicision based on topologyKey and other constraints like cpu request, then bound the PV and PVC.  Not sure if I understand it correctly, but couple of question off my head regarding this workflow:

- How PV/PVC controller knows NOT to bind the PVC? There should be a way to differentiate `local` and `remote`. I believe this is discussed in the proposal, but I don't see any API defined here, do we need another field for the purpose?
- PV is created by discovery daemon, how does it know what storageClassName for PVs it created, do we need admin to configure it?
- PV and PVC both have storageClassName, is this required? am i missing something?